src/Storages/FileLog/FileLogSource.cpp

```diff
@@ -0,0 +1,126 @@
+#include <DataStreams/ConvertingBlockInputStream.h>
+#include <DataStreams/OneBlockInputStream.h>
+#include <Formats/FormatFactory.h>
+#include <Interpreters/Context.h>
+#include <Processors/Executors/StreamingFormatExecutor.h>
+#include <Processors/Formats/InputStreamFromInputFormat.h>
+#include <Storages/FileLog/FileLogSource.h>
+#include <Storages/FileLog/ReadBufferFromFileLog.h>
+#include <Common/Stopwatch.h>
+#include <base/logger_useful.h>
+
+namespace DB
+{
+static constexpr auto MAX_FAILED_POLL_ATTEMPTS = 10;
+
+FileLogSource::FileLogSource(
+    StorageFileLog & storage_,
+    const StorageMetadataPtr & metadata_snapshot_,
+    const ContextPtr & context_,
+    const Names & columns,
+    size_t max_block_size_,
+    size_t poll_time_out_,
+    size_t stream_number_,
+    size_t max_streams_number_)
+    : SourceWithProgress(metadata_snapshot_->getSampleBlockForColumns(columns, storage_.getVirtuals(), storage_.getStorageID()))
+    , storage(storage_)
+    , metadata_snapshot(metadata_snapshot_)
+    , context(context_)
+    , column_names(columns)
+    , max_block_size(max_block_size_)
+    , poll_time_out(poll_time_out_)
+    , stream_number(stream_number_)
+    , max_streams_number(max_streams_number_)
+    , non_virtual_header(metadata_snapshot_->getSampleBlockNonMaterialized())
+    , virtual_header(
+          metadata_snapshot->getSampleBlockForColumns(storage.getVirtualColumnNames(), storage.getVirtuals(), storage.getStorageID()))
+{
+    buffer = std::make_unique<ReadBufferFromFileLog>(storage, max_block_size, poll_time_out, context, stream_number_, max_streams_number_);
+
+    const auto & file_infos = storage.getFileInfos();
+
+    size_t files_per_stream = file_infos.file_names.size() / max_streams_number;
+    start = stream_number * files_per_stream;
+    end = stream_number == max_streams_number - 1 ? file_infos.file_names.size() : (stream_number + 1) * files_per_stream;
+}
+
+void FileLogSource::onFinish()
+{
+    storage.closeFilesAndStoreMeta(start, end);
+}
+
+Chunk FileLogSource::generate()
+{
+    if (!buffer || buffer->noRecords())
+    {
+        /// There is no onFinish for ISource, we call it
+        /// when no records return to close files
+        onFinish();
+        return {};
```

Comment:
But will onFinish be executed if we exit in line 104?
may after all makes sence to also move it to cdesstructor -- under `if (!finished)` flag.